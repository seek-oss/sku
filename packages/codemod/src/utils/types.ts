export type Transform = (source: string) => Promise<string | null>;

export const isValidTransform = (fn: unknown): fn is Transform =>
  typeof fn === 'function';
