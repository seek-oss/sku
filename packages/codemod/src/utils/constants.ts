type Codemod = {
  description: string;
  value: string;
};

export const CODEMODS = [
  {
    description:
      'Transform all webpack loadable imports to Vite loadable imports',
    value: 'transform-vite-loadable',
  },
  {
    description: 'Convert Jest tests to Vitest',
    value: 'jest-to-vitest',
  },
  /* [add-sku-codemod-generator: codemod] */
] as const satisfies Codemod[];

export type CodemodName = (typeof CODEMODS)[number]['value'];
