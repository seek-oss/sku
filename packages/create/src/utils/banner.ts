import chalk from 'chalk';

export const banner = (
  type: 'info' | 'success' | 'warning' | 'error',
  title: string,
  messages: string[] = [],
) => {
  const colorMap = {
    info: chalk.blue,
    success: chalk.green,
    warning: chalk.yellow,
    error: chalk.red,
  };

  const color = colorMap[type];
  let icon = 'ℹ️';
  if (type === 'success') {
    icon = '✅';
  } else if (type === 'error') {
    icon = '❌';
  }

  console.log();
  console.log(color(`${icon} ${title}`));

  if (messages.length > 0) {
    console.log();
    messages.forEach((message) => {
      console.log(`  ${message}`);
    });
  }

  console.log();
};
