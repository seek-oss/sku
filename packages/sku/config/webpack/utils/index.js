const loaders = require('./loaders');
const { resolvePackage } = require('./resolvePackage');

module.exports = {
  ...loaders,
  resolvePackage,
  TYPESCRIPT: /\.(ts|tsx)$/,
  JAVASCRIPT: /\.js$/,
  LESS: /\.less$/,
  IMAGE: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
  SVG: /\.svg$/,
};
