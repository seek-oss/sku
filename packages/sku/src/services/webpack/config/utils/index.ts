import { extensions } from '../../../../config/fileResolutionExtensions.js';

export * from './loaders.js';
export { resolvePackage } from './resolvePackage.js';

export const TYPESCRIPT = new RegExp(`\.(${extensions.ts.join('|')})$`);
export const JAVASCRIPT = new RegExp(`\.(${extensions.js.join('|')})$`);
export const IMAGE = [
  /\.bmp$/,
  /\.gif$/,
  /\.jpe?g$/,
  /\.png$/,
  /\.webp$/,
  /\.avif$/,
];
export const SVG = /\.svg$/;
