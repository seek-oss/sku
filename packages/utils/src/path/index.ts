export { toPosixPath } from './toPosixPath.js';
