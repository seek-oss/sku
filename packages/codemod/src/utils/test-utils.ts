import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';
import { isValidTransform, type Transform } from './types.js';

type FileInfo = {
  source: string;
  path: string;
};

type InlineTestProps = {
  transform: Transform;
  input: FileInfo;
  expectedOutput: string;
};

type TestProps = {
  dirName: string;
  transformName: string;
  fixtureName?: string;
  testOptions?: { extension?: string };
};

const runInlineTest = async ({
  transform,
  input,
  expectedOutput,
}: InlineTestProps) => {
  if (!isValidTransform(transform)) {
    throw new Error('Invalid transform function provided');
  }

  const output = transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }

  expect(output.trim()).toEqual(expectedOutput.trim());

  return output;
};

export const runTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const testFixtureName = fixtureName || transformName;

  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${testFixtureName}.input.${testOptions.extension}`,
  );

  // Assumes transform is one level up from __tests__ directory
  const transformImport = import(path.join(dirName, '..', transformName));
  const sourceRead = readFile(inputPath, 'utf8');
  const expectedOutputRead = readFile(
    path.join(fixtureDir, `${testFixtureName}.output.${testOptions.extension}`),
    'utf8',
  );

  const [{ transform }, source, expectedOutput] = await Promise.all([
    transformImport,
    sourceRead,
    expectedOutputRead,
  ]);

  return runInlineTest({
    transform,
    input: {
      path: inputPath,
      source,
    },
    expectedOutput,
  });
};

export const runNoChangeTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${fixtureName}.unchanged.${testOptions.extension}`,
  );

  const sourceRead = readFile(inputPath, 'utf8');
  const transformImport = import(path.join(dirName, '..', transformName));

  const [{ transform }, source] = await Promise.all([
    transformImport,
    sourceRead,
  ]);
  const output = await transform(source);

  expect(output).toEqual(null);
};
