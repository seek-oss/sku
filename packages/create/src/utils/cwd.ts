import { join } from 'node:path';

let cwd = process.cwd();

export const setCwd = (dir: string) => {
  cwd = dir;
};

export const getCwd = () => cwd;

export const getPathFromCwd = (filePath: string) => join(cwd, filePath);
