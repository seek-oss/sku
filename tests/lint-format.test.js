const path = require('node:path');
const fs = require('node:fs/promises');
const dedent = require('dedent');
const { runSkuScriptInDir } = require('@sku-private/test-utils');

const appDirectory = path.dirname(
  require.resolve('@sku-fixtures/lint-format/sku.config.ts'),
);
const srcDirectory = path.join(appDirectory, 'src');
const testFile = (fileName) => path.join(srcDirectory, fileName);
const stripAnsi = require('strip-ansi');

const filesToLint = {
  'utils.test.ts': dedent/* ts */ `
    console.log('foo');

    it.only('should test something', () => {
      let foo = true;

      expect(foo).toBe(true);
    });\n`,
  'typescriptFile.ts': dedent/* ts */ `
    const foo: number = 'a string';
  `,
  'unformattedFile1.js': dedent/* js */ `
    import { something } from "with-double-quotes";

      const indented = 'indented';

    const foo = [
      "something \"really\" long",
      "that has to be moved",
      'to multiple lines',
      "so we can 'test' trailing commas"
    ];
  `,
};

const filesToFormat = {
  'importOrder1.ts': dedent/* ts */ `
    import './reset'; // side-effect imports should stay put

    import LocalComponent from './LocalComponent'; // sibling
    import Parent from '../parent'; // parents
    import someModule from 'some-module'; // external
    import vanillaStyles from './vanillaStyles.css'; // styles
    import distantParent from '../../../parent'; // parents
    import myself from '.'; // index
    import path from 'path'; //  built-in
    import styles from './styles.less'; // styles
    import utils from 'src/utils'; // internal
  `,
  'importOrder2.ts': dedent/* ts */ `
    import aStyle from './a.css';
    import bStyle from './b.css';
    import cStyle from '../c.css';

    import b from './b';
    import a from './a';
    import c from '../c';
  `,
  'fixableLintError.ts': dedent/* ts */ `
    const foo = () => {
      return 'foo';
    };
  `,
  'unformattedFile2.ts': dedent/* ts */ `
    import { something } from "with-double-quotes";

      const indented = 'indented';

    const foo = [
      "something \"really\" long",
      "that has to be moved",
      'to multiple lines',
      "so we can 'test' trailing commas"
    ];
  `,
};

beforeEach(async () => {
  await fs.mkdir(srcDirectory, { recursive: true });
});

afterEach(async () => {
  await fs.rm(srcDirectory, { recursive: true, force: true });
});

expect.addSnapshotSerializer({
  serialize: (val) => {
    const { stdout } = val;
    // Remove some logs that contain file paths that are unique to the machine
    const sanitizedStdout = stdout
      .split('\n')
      .filter((line) => !line.includes('sku/fixtures/lint-format'))
      .join('\n');

    return dedent`
      stdout: ${stripAnsi(sanitizedStdout)}
    `;
  },
  test: (val) => typeof val === 'object' && val.hasOwnProperty('stdout'),
});

describe('sku lint', () => {
  test.each(Object.keys(filesToLint))(
    'lint errors are reported: %s',
    async (fileName) => {
      const filePath = testFile(fileName);
      await fs.writeFile(filePath, filesToLint[fileName]);

      let result;

      try {
        await runSkuScriptInDir('lint', appDirectory);
      } catch (err) {
        result = { stderr: err.stderr, stdout: err.stdout };
      }

      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    },
  );
});

describe('sku format', () => {
  expect.addSnapshotSerializer({
    serialize: (val) => val,
    test: (val) => typeof val === 'string',
  });

  test.each(Object.keys(filesToFormat))(
    'errors are fixed: %s',
    async (fileName) => {
      const filePath = testFile(fileName);
      await fs.writeFile(filePath, filesToFormat[fileName]);

      await runSkuScriptInDir('format', appDirectory);

      const result = await fs.readFile(filePath, { encoding: 'utf-8' });

      expect(result).toMatchSnapshot();
    },
  );
});
