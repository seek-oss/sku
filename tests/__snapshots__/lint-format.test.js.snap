// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lint-format > sku format > errors are fixed: fixableLintError.ts 1`] = `
const foo = () => 'foo';

`;

exports[`lint-format > sku format > errors are fixed: importOrder1.ts 1`] = `
import './reset'; // side-effect imports should stay put

import path from 'path'; //  built-in

import someModule from 'some-module'; // external

import utils from 'src/utils'; // internal

import distantParent from '../../../parent'; // parents
import Parent from '../parent'; // parents

import LocalComponent from './LocalComponent'; // sibling

import myself from '.'; // index

import vanillaStyles from './vanillaStyles.css'; // styles

`;

exports[`lint-format > sku format > errors are fixed: importOrder2.ts 1`] = `
import c from '../c';

import a from './a';
import b from './b';

import aStyle from './a.css';
import bStyle from './b.css';
import cStyle from '../c.css';

`;

exports[`lint-format > sku format > errors are fixed: unformattedFile2.ts 1`] = `
import { something } from 'with-double-quotes';

const indented = 'indented';

const foo = [
  'something "really" long',
  'that has to be moved',
  'to multiple lines',
  "so we can 'test' trailing commas",
];

`;

exports[`lint-format > sku lint > lint errors are reported: typescriptFile.ts 1`] = `
stdout: Linting
Checking code with TypeScript compiler
src/importOrder1.ts(3,28): error TS2307: Cannot find module './LocalComponent' or its corresponding type declarations.
src/importOrder1.ts(4,20): error TS2307: Cannot find module '../parent' or its corresponding type declarations.
src/importOrder1.ts(5,24): error TS2307: Cannot find module 'some-module' or its corresponding type declarations.
src/importOrder1.ts(6,27): error TS2307: Cannot find module './vanillaStyles.css' or its corresponding type declarations.
src/importOrder1.ts(7,27): error TS2307: Cannot find module '../../../parent' or its corresponding type declarations.
src/importOrder1.ts(8,20): error TS2307: Cannot find module '.' or its corresponding type declarations.
src/importOrder1.ts(10,19): error TS2307: Cannot find module 'src/utils' or its corresponding type declarations.
src/importOrder2.ts(1,20): error TS2307: Cannot find module './a.css' or its corresponding type declarations.
src/importOrder2.ts(2,20): error TS2307: Cannot find module './b.css' or its corresponding type declarations.
src/importOrder2.ts(3,20): error TS2307: Cannot find module '../c.css' or its corresponding type declarations.
src/importOrder2.ts(5,15): error TS2307: Cannot find module './b' or its corresponding type declarations.
src/importOrder2.ts(6,15): error TS2307: Cannot find module './a' or its corresponding type declarations.
src/importOrder2.ts(7,15): error TS2307: Cannot find module '../c' or its corresponding type declarations.
src/typescriptFile.ts(1,7): error TS2322: Type 'string' is not assignable to type 'number'.
src/utils.test.ts(3,1): error TS2582: Cannot find name 'it'. Do you need to install type definitions for a test runner? Try \`npm i --save-dev @types/jest\` or \`npm i --save-dev @types/mocha\`.
src/utils.test.ts(6,3): error TS2304: Cannot find name 'expect'.
`;

exports[`lint-format > sku lint > lint errors are reported: unformattedFile1.js 1`] = `
stdout: Linting
Checking code with TypeScript compiler
src/importOrder1.ts(3,28): error TS2307: Cannot find module './LocalComponent' or its corresponding type declarations.
src/importOrder1.ts(4,20): error TS2307: Cannot find module '../parent' or its corresponding type declarations.
src/importOrder1.ts(5,24): error TS2307: Cannot find module 'some-module' or its corresponding type declarations.
src/importOrder1.ts(6,27): error TS2307: Cannot find module './vanillaStyles.css' or its corresponding type declarations.
src/importOrder1.ts(7,27): error TS2307: Cannot find module '../../../parent' or its corresponding type declarations.
src/importOrder1.ts(8,20): error TS2307: Cannot find module '.' or its corresponding type declarations.
src/importOrder1.ts(10,19): error TS2307: Cannot find module 'src/utils' or its corresponding type declarations.
src/importOrder2.ts(1,20): error TS2307: Cannot find module './a.css' or its corresponding type declarations.
src/importOrder2.ts(2,20): error TS2307: Cannot find module './b.css' or its corresponding type declarations.
src/importOrder2.ts(3,20): error TS2307: Cannot find module '../c.css' or its corresponding type declarations.
src/importOrder2.ts(5,15): error TS2307: Cannot find module './b' or its corresponding type declarations.
src/importOrder2.ts(6,15): error TS2307: Cannot find module './a' or its corresponding type declarations.
src/importOrder2.ts(7,15): error TS2307: Cannot find module '../c' or its corresponding type declarations.
src/typescriptFile.ts(1,7): error TS2322: Type 'string' is not assignable to type 'number'.
src/utils.test.ts(3,1): error TS2582: Cannot find name 'it'. Do you need to install type definitions for a test runner? Try \`npm i --save-dev @types/jest\` or \`npm i --save-dev @types/mocha\`.
src/utils.test.ts(6,3): error TS2304: Cannot find name 'expect'.
`;

exports[`lint-format > sku lint > lint errors are reported: utils.test.ts 1`] = `
stdout: Linting
Checking code with TypeScript compiler
src/importOrder1.ts(3,28): error TS2307: Cannot find module './LocalComponent' or its corresponding type declarations.
src/importOrder1.ts(4,20): error TS2307: Cannot find module '../parent' or its corresponding type declarations.
src/importOrder1.ts(5,24): error TS2307: Cannot find module 'some-module' or its corresponding type declarations.
src/importOrder1.ts(6,27): error TS2307: Cannot find module './vanillaStyles.css' or its corresponding type declarations.
src/importOrder1.ts(7,27): error TS2307: Cannot find module '../../../parent' or its corresponding type declarations.
src/importOrder1.ts(8,20): error TS2307: Cannot find module '.' or its corresponding type declarations.
src/importOrder1.ts(10,19): error TS2307: Cannot find module 'src/utils' or its corresponding type declarations.
src/importOrder2.ts(1,20): error TS2307: Cannot find module './a.css' or its corresponding type declarations.
src/importOrder2.ts(2,20): error TS2307: Cannot find module './b.css' or its corresponding type declarations.
src/importOrder2.ts(3,20): error TS2307: Cannot find module '../c.css' or its corresponding type declarations.
src/importOrder2.ts(5,15): error TS2307: Cannot find module './b' or its corresponding type declarations.
src/importOrder2.ts(6,15): error TS2307: Cannot find module './a' or its corresponding type declarations.
src/importOrder2.ts(7,15): error TS2307: Cannot find module '../c' or its corresponding type declarations.
src/typescriptFile.ts(1,7): error TS2322: Type 'string' is not assignable to type 'number'.
src/utils.test.ts(3,1): error TS2582: Cannot find name 'it'. Do you need to install type definitions for a test runner? Try \`npm i --save-dev @types/jest\` or \`npm i --save-dev @types/mocha\`.
src/utils.test.ts(6,3): error TS2304: Cannot find name 'expect'.
`;

exports[`sku format > errors are fixed: fixableLintError.ts 1`] = `
const foo = () => 'foo';

`;

exports[`sku format > errors are fixed: importOrder1.ts 1`] = `
import './reset'; // side-effect imports should stay put

import path from 'path'; //  built-in

import someModule from 'some-module'; // external

import utils from 'src/utils'; // internal

import distantParent from '../../../parent'; // parents
import Parent from '../parent'; // parents

import LocalComponent from './LocalComponent'; // sibling

import myself from '.'; // index

import vanillaStyles from './vanillaStyles.css'; // styles

`;

exports[`sku format > errors are fixed: importOrder2.ts 1`] = `
import c from '../c';

import a from './a';
import b from './b';

import aStyle from './a.css';
import bStyle from './b.css';
import cStyle from '../c.css';

`;

exports[`sku format > errors are fixed: unformattedFile2.ts 1`] = `
import { something } from 'with-double-quotes';

const indented = 'indented';

const foo = [
  'something "really" long',
  'that has to be moved',
  'to multiple lines',
  "so we can 'test' trailing commas",
];

`;

exports[`sku lint > lint errors are reported: typescriptFile.ts 1`] = `
stdout: Linting
Checking code with TypeScript compiler
src/typescriptFile.ts(1,7): error TS2322: Type 'string' is not assignable to type 'number'.
`;

exports[`sku lint > lint errors are reported: unformattedFile1.js 1`] = `
stdout: Linting
Checking code with TypeScript compiler
Checking code with Prettier
Paths: .
src/unformattedFile1.js
To fix this issue, run 'pnpm run format'
`;

exports[`sku lint > lint errors are reported: utils.test.ts 1`] = `
stdout: Linting
Checking code with TypeScript compiler
Checking code with Prettier
Paths: .
Checking code with ESLint
Paths: .

1:1  error  Unexpected console statement                    no-console
3:4  error  Unexpected focused test                         jest/no-focused-tests
4:7  error  'foo' is never reassigned. Use 'const' instead  prefer-const

âœ– 3 problems (3 errors, 0 warnings)
1 error and 0 warnings potentially fixable with the \`--fix\` option.
`;
